### PR TITLE
Activate Tabs by hovering an element over them

### PR DESCRIPTION
**Pull request by fabian-bizfactory**
This is implemented for:
- The big Tab Navigation
- Small Tabs

### CODE REVIEW

File: frontend/src/js/small-tab-navigation/SmallTabNavigation.tsx [hunk 1]
```diff
@@ -26,7 +26,7 @@ const SmallTabNavigation: FC<PropsT> = ({
   onSelectTab,
 }) => {
   return (
-    <div className={className}>
+    <div className={className} style={{"display": "flex"}}>
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/SmallTabNavigation.tsx:29]: What's the reason for that change? The parent components are currently setting layout settings like this. Plus, if we style, let's please try and avoid inline styles and use styled components instead

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// 600 ms is the default time for a tooltip to appear
+// feels very responsive
+// can be overriden by passing a timeUntilNavigateOverride prop
+const TIME_UNTIL_NAVIGATE = 600;
+
+export const HoverNagiatable:FC<PropsT> = ({ timeUntilNavigateOverride, triggerNavigate, children, disabled }) => {
+    const time_to_navigate = timeUntilNavigateOverride || TIME_UNTIL_NAVIGATE;
+    let timeout: number | null | NodeJS.Timeout; // timeout can either be a number or a NodeJS.Timeout depending on the environment
+    const [_, drop] = useDrop({
+        accept: [DNDType.FORM_CONFIG, DNDType.CONCEPT_TREE_NODE, DNDType.PREVIOUS_QUERY, DNDType.PREVIOUS_SECONDARY_ID_QUERY],
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:22]: Configuring prettier will also format those lines

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+// default time until the hover triggers navigation
+// 600 ms is the default time for a tooltip to appear
+// feels very responsive
+// can be overriden by passing a timeUntilNavigateOverride prop
+const TIME_UNTIL_NAVIGATE = 600;
+
+export const HoverNagiatable:FC<PropsT> = ({ timeUntilNavigateOverride, triggerNavigate, children, disabled }) => {
+    const time_to_navigate = timeUntilNavigateOverride || TIME_UNTIL_NAVIGATE;
+    let timeout: number | null | NodeJS.Timeout; // timeout can either be a number or a NodeJS.Timeout depending on the environment
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:20]: why did you add that comment? Depending on which environment? browser vs server? We're only running the app in the browser

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { FC } from "react";
+import { useDrop } from "react-dnd";
+import { DNDType } from "../common/constants/dndTypes";
+
+
+interface PropsT {
+    timeUntilNavigateOverride?: number;
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:7]: Let's not add props that we don't use. Even though it's a nice thought for the future, if we're never passing `timeUntilNavigateOverride` anywhere, let's remove the prop

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+const TIME_UNTIL_NAVIGATE = 600;
+
+export const HoverNagiatable:FC<PropsT> = ({ timeUntilNavigateOverride, triggerNavigate, children, disabled }) => {
+    const time_to_navigate = timeUntilNavigateOverride || TIME_UNTIL_NAVIGATE;
+    let timeout: number | null | NodeJS.Timeout; // timeout can either be a number or a NodeJS.Timeout depending on the environment
+    const [_, drop] = useDrop({
+        accept: [DNDType.FORM_CONFIG, DNDType.CONCEPT_TREE_NODE, DNDType.PREVIOUS_QUERY, DNDType.PREVIOUS_SECONDARY_ID_QUERY],
+        hover: (_, monitor) => {
+            // uses property of null => if null == false
+            if (timeout == null && !disabled) {
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:25]: What's the reason for a non-explicit equality check? Let's please use explicit equality checks, so `===` instead. (we should configure this as an eslint rule)

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+// can be overriden by passing a timeUntilNavigateOverride prop
+const TIME_UNTIL_NAVIGATE = 600;
+
+export const HoverNagiatable:FC<PropsT> = ({ timeUntilNavigateOverride, triggerNavigate, children, disabled }) => {
+    const time_to_navigate = timeUntilNavigateOverride || TIME_UNTIL_NAVIGATE;
+    let timeout: number | null | NodeJS.Timeout; // timeout can either be a number or a NodeJS.Timeout depending on the environment
+    const [_, drop] = useDrop({
+        accept: [DNDType.FORM_CONFIG, DNDType.CONCEPT_TREE_NODE, DNDType.PREVIOUS_QUERY, DNDType.PREVIOUS_SECONDARY_ID_QUERY],
+        hover: (_, monitor) => {
+            // uses property of null => if null == false
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:24]: What do you mean by that comment? Could you clarify this?

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import { FC } from "react";
+import { useDrop } from "react-dnd";
+import { DNDType } from "../common/constants/dndTypes";
+
+
+interface PropsT {
+    timeUntilNavigateOverride?: number;
+    triggerNavigate: () => void;
+    disabled?: boolean;
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:9]: Same for `disabled`

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx (hunk 3 above)
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:20]: What happens when the component rerenders? Will the timeout just get overridden? I think we should probably use a timeout ref instead of a local variable

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+// default time until the hover triggers navigation
+// 600 ms is the default time for a tooltip to appear
+// feels very responsive
+// can be overriden by passing a timeUntilNavigateOverride prop
+const TIME_UNTIL_NAVIGATE = 600;
+
+export const HoverNagiatable:FC<PropsT> = ({ timeUntilNavigateOverride, triggerNavigate, children, disabled }) => {
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:18]: Going forward, let's try and avoid React.FC, see https://github.com/facebook/create-react-app/pull/8177. Preferrably, let's just inline props definitions like `const Comp = ({ a, b }: { a: string, b: number}) => {`

**Review by Kadrian — CHANGES_REQUESTED** [review]
Alright, so far so good. I think the approach to use a `useDrop` hover + `monitor.isOver()` is good. I added some comments.

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx (hunk 3 above)
Comment (fabian-bizfactory) [reply frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:20]: There are two reasons, why I added the NodeJS.Timeout (sorry for the misleading comment)
- Typescript does not like it if the Type is missing

File: frontend/src/js/small-tab-navigation/SmallTabNavigation.tsx (hunk 1 above)
Comment (fabian-bizfactory) [reply frontend/src/js/small-tab-navigation/SmallTabNavigation.tsx:29]: image
Breaks due to the Div in HoverNavigatable. But I will remove the style here

File: frontend/src/js/standard-query-editor/QueryNodeActions.tsx [hunk 9]
```diff
@@ -18,7 +18,7 @@ const StyledFaIcon = styled(FaIcon)`
 `;
 
 const StyledIconButton = styled(IconButton)`
-  padding: 0px 6px 4px;
+  padding: 3.5px 6px 0px;
```
Comment (Kadrian) [inline frontend/src/js/standard-query-editor/QueryNodeActions.tsx:21]: Hmm, how is this change related? Let's please avoid half-pixel values

**Review by Kadrian — CHANGES_REQUESTED** [review]
When I tested this
- I can't see a hover indication when hovering over a tab, shouldn't there be one?
- the tab switch stops working after the switch happened once. So I can't hover Editor -> Form Editor -> Back to Editor. Could you look into that?

File: frontend/src/js/standard-query-editor/QueryNodeActions.tsx (hunk 9 above)
Comment (fabian-bizfactory) [reply frontend/src/js/standard-query-editor/QueryNodeActions.tsx:21]: This was to fix 
image

**Comment by Kadrian** [comment]
Could you please check all your changes and fix the code formatting?

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+const Root = styled("div")<{
+  isOver?: boolean;
+}>`
+  border: ${({ theme, isOver }) =>
+    isOver ? `1px solid ${theme.col.grayMediumLight}` : "none"};
+  position: relative;
+  display: inline-flex;
+  color: ${({ theme, isOver }) =>
+    isOver ? theme.col.red : isOver ? theme.col.black : theme.col.gray};
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:20]: What does this logic do, can you explain?

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+`;
+
+// default time until the hover triggers navigation
+// 600 ms is the default time for a tooltip to appear
+// feels very responsive
+// can be overriden by passing a timeUntilNavigateOverride prop
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:26]: That note can be removed, right, since there's no  timeUntilNavigateOverride?

File: frontend/src/js/small-tab-navigation/SmallTabNavigationButton.tsx [hunk 12]
```diff
@@ -76,16 +77,19 @@ const SmallTabNavigationButton = forwardRef<HTMLButtonElement, PropsT>(
...
     return (
-      <Button
-        ref={ref}
-        highlightColor={highlightColor}
-        type="button"
-        size={size}
-        selected={isSelected}
-        onClick={onClick}
+      <HoverNavigatable triggerNavigate={onClick}
       >
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/SmallTabNavigationButton.tsx:81]: Still not formatted, is it?

File: frontend/src/js/small-tab-navigation/SmallTabNavigationButton.tsx [hunk 13]
```diff
@@ -10,7 +11,7 @@ const bottomBorderBase = css`
   height: 3px;
   bottom: 0px;
 `;
-const Button = styled("button")<{
+const Button = styled("button") <{
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/SmallTabNavigationButton.tsx:14]: Here as well

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  });
+  return (
+    <Root ref={drop} isOver={isOver}>
+      {children}
+    </Root>
+  );
+};
+
+export const FlexHoverNavigatable = styled(HoverNavigatable)`
+  flex: flex;
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:63]: What is `flex: flex`?

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    }),
+  });
+  return (
+    <Root ref={drop} isOver={isOver}>
+      {children}
+    </Root>
+  );
+};
+
+export const FlexHoverNavigatable = styled(HoverNavigatable)`
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:62]: Let's have calling components re-style `HoverNavigatable` if needed

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+interface PropsT {
+  triggerNavigate: () => void;
+  children: ReactNode;
+}
+
+const Root = styled("div")<{
+  isOver?: boolean;
+}>`
+  border: ${({ theme, isOver }) =>
+    isOver ? `1px solid ${theme.col.grayMediumLight}` : "none"};
```
Comment (Kadrian) [inline frontend/src/js/small-tab-navigation/HoverNavigatable.tsx:16]: Using a border for highlight makes the rest of the layout jump. How about using a light gray background color instead?

**Review by Kadrian — CHANGES_REQUESTED** [review]

File: frontend/src/js/small-tab-navigation/HoverNavigatable.tsx (hunk 16 above)
Comment (fabian-bizfactory) [reply frontend/src/js/small-tab-navigation/HoverNavigatable.tsx]: I have just implemented that using very light gray background. Please let me know what you think

File: frontend/src/js/standard-query-editor/QueryNode.tsx [hunk 17]
```diff
@@ -127,49 +133,51 @@ const QueryNode = ({
...
+        description: node.description,
+        tree: node.tree,
+        tables: node.tables,
+        selects: node.selects,
+
+        additionalInfos: node.additionalInfos,
+        matchingEntries: node.matchingEntries,
+        matchingEntities: node.matchingEntities,
+        dateRange: node.dateRange,
+      }
```
Comment (Kadrian) [inline frontend/src/js/standard-query-editor/QueryNode.tsx:147]: I added a [CI step for lint & checking format](https://github.com/bakdata/conquery/pull/2856). Could you please merge develop into this branch and fix the formatting issues until the pipeline is green?

File: frontend/src/js/standard-query-editor/QueryNode.tsx [hunk 18]
```diff
@@ -180,37 +187,39 @@ const QueryNode = ({
...
-        excludeFromSecondaryId={node.excludeFromSecondaryId}
-        onDeleteNode={onDeleteNode}
-        onToggleTimestamps={onToggleTimestamps}
-        onToggleSecondaryIdExclude={onToggleSecondaryIdExclude}
-        onExpandClick={expandClick}
-        previousQueryLoading={node.loading}
-        error={node.error}
-      />
-    </Root>
+    <FlexHoverNavigatable triggerNavigate={onClick}>
```
Comment (Kadrian) [inline frontend/src/js/standard-query-editor/QueryNode.tsx:190]: If we're adding this to `QueryNode`, let's also add it to the concepts in the forms, so to `FormConceptNode`

**Review by Kadrian — APPROVED** [review]
Nice, I tested this again, feels good now with the background color change. Code also looks good to me now.

Right now the feature is still missing in `FormConceptNode` – feel free to either add this in here or open a follow up PR
